deploy/contracts/AddressProxy.sol

```diff
@@ -0,0 +1,157 @@
+pragma solidity ^0.4.19;
+
+
+contract AddressProxy {
+
+    struct ContractAddress {
+        uint id;
+        string name;
+        address at;
+        uint createdTime;
+        uint updatedTime;
+    }
+
+    address public owner;
+    mapping(uint => ContractAddress) private contractsAddress;
+    uint[] public availableIds;
```

Comment:
What's the necessity of ids here? Couldn't we just store a list of contract names and the mapping as (bytes => ContractAddress)? It might be more gas efficient

Reply:
@tcrosado I opted for this solution because in Solidiy you cannot retrieve the available keys of a Map.

The AddressProxy need to accept more addresses in the future (AppCoinsBClass, the Storage Contracts) so having an array that we can use to find the available keys is useful.

Reply:
Maybe i didn't express myself in the best way. I agree with you, i'm suggesting using the contract name as key and keep also a list of keys instead of ids. Because right now we are not taking advantage of the hashmap to immediately find the contract address as you have to iterate over the avaliableIds to then match with the contract name.

Reply:
Just to make sure I understood the contract's functionality correctly:
- There's a list of integer IDs that represent the available contracts
- There's a mapping between those IDs and the contracts
- There are methods to directly change the addresses of the contracts we have today
- We can add addresses for other contracts that we may create but they won't have their own "direct methods" (methods used to update their addresses directly) unless we change **AddressProxy**, which is something we want to avoid

So I have 2 questions:
1. Why should we have the "direct methods" for the current contracts if then **AddressProxy** won't have this kind of methods for other contracts we create in the future?
2. As @tcrosado said, can't we have the names of the contracts as their IDs instead of integers? I think it would be easier and cheaper to manage the access to a contract's address since we can get the available names and then just access the mapping with that name as key. Right now, if we get the available IDs, we don't know the mapping between ID and contract's name.

Reply:
ok, i will try answer both at the same time:

Reply:
I will try to answer both at the same time:

Reply:
ok, let me try for the second time after loosing the reply, i will answer both at the same time:


#### Why should we have the "direct methods" for the current contracts if then AddressProxy won't have this kind of methods for other contracts we create in the future?
- This is true, i will remove those.


#### Maybe i didn't express myself in the best way. I agree with you, i'm suggesting using the contract name as key and keep also a list of keys instead of ids. Because right now we are not taking advantage of the hashmap to immediately find the contract address as you have to iterate over the avaliableIds to then match with the contract name. As it's done on Advertisement contract in the mapping (country => campaignIdList).
####  Maybe i didn't express myself in the best way. I agree with you, i'm suggesting using the contract name as key and keep also a list of keys instead of ids. Because right now we are not taking advantage of the hashmap to immediately find the contract address as you have to iterate over the avaliableIds to then match with the contract name. As it's done on Advertisement contract in the mapping (country => campaignIdList).

- The reason i went for the mapping of integer (the map there was the same as an array) was to make it more human readable, i will explain by though proccess even though i think string => obj is better:
- In soldity you cannot return an array of strings so you are confined to returning an array of bytes32, and since an array of bytes32 has the same as intengers in terms of process, it is easier to use an integer.

Can we agree on using this?:

```
    mapping(string => ContractAddress) private contractsAddress;
    uint[] public availableIds;
```

I think i tried to much to make it as easy to use as possible without considering those implications.

Reply:
> Can we agree on using this?
```
mapping(string => ContractAddress) private contractsAddress;
bytes32[] public availableIds;
```

Yes, I agree with this approach. Then the pattern of accessing the address of a *ContractAddress* is:
1. calling *getAvailableIds()*
2. calling *getContractAddress(string name)*

With this change, we only need to keep the following methods right?
- addAddress
- getAvailableIds
- getContractAddress (method that receives a name and outputs the address)
- getContractCreatedTime
- getContractUpdatedTime

These last 2 methods don't exist yet but if the *ContractAddress* structure has these attributes, there should be methods to access them.

Reply:
@DioPires The approach should be:

Reply:
@DioPires the flow shoud be:

Reply:
@DioPires the approach should be:

1. Get getAvailableIds() 
//=> ["0x1262","0x1245","0x1236"]

2. Get getContractAddressById(0x1262)
//=> address 0xc16892b53362dce249bba139bbe913b548fcd233


Since solidity does not return a array of stirngs, we need to convert them to bytes32.
So the available *public* methods are:

- addAddress
- getContractNameById (this one helps you convert from bytes32 to string)
- getContractAddressById
- getContractCreatedTimeById
- getContractUpdatedTimeById
- getContractAddressByName <- do you think you should have this one? If the user want to use a string instead of an bytes32?

Reply:
> @DioPires the approach should be:
Get getAvailableIds()
//=> ["0x1262","0x1245","0x1236"]
Get getContractAddressById(0x1262)
//=> address 0xc16892b53362dce249bba139bbe913b548fcd233

You're right.

> getContractAddressByName <- do you think you should have this one? If the user want to use a string instead of an bytes32?
I don't think we need it, because the client would need to know the contract names beforehand. What do you think @tcrosado?

Reply:
> I don't think we need it, because the client would need to know the contract names beforehand. 

Yes, Then we should make the mapping to  mapping(bytes32 => ContractAddress) private contractsAddress;

Updating the contract.

Reply:
Ok I understand. There is no need for a contract name list.